### PR TITLE
Alternative/extended approach to self-references

### PR DESCRIPTION
**Pull request by webpro**
After your pull request about the same, today I went pretty deep into the self-referencing machinery.

The two main points I tried to address in this PR:

1. What I meant with https://github.com/webpro/knip/pull/88#discussion_r1137154890 was to literally use only that condition, without needing to pass `workspace` into this AST walker. Then after `getImportsAndExports` is done this `entryPath` needs to be added as such since the self-referencing import may refer to regular source code (not necessarily built into `/dist` or something).
2. We can also `createRequire`  "on demand" in the main flow with having to pass the `require` and `resolve` around with the workspace. Since this isn't used that often think it's acceptable and keeps the code a bit more clean (also the debug output). Could cache `require` instances if really necessary, but I'm guessing Node.js already takes care of that.

I also added the `workspaces-self-reference` test and fixture to assert the self-references and cross-references are working in workspaces as well.

Quite the ride. I hope you can follow along. What do you think? Sorry if it's hard to follow. Happy to discuss further. And if you think "if it works it's I'm fine with it" that's cool too, we can just merge and move on.

### CODE REVIEW

**Comment by webpro** [comment]
Definitely not proud of it, but I added a commit to suppress issues for Windows. I prefer not to wait for upstream fixes for this specific issue.

**Comment by webpro** [comment]
Windows issue should be fixed now (https://github.com/webpro/knip/actions/runs/4442355797)

**Comment by RebeccaStevens** [comment]
Ahh, I missed this. Giving it a quick look through, I like it :)
